runtime/src/bank.rs

```diff
@@ -1311,6 +1327,47 @@ impl Bank {
         self.update_recent_blockhashes_locked(&blockhash_queue);
     }
 
+    fn get_timestamp_estimate(&self) -> Option<UnixTimestamp> {
+        let mut get_timestamp_estimate_time = Measure::start("get_timestamp_estimate");
+        let recent_timestamps: HashMap<Pubkey, (Slot, UnixTimestamp)> = self
+            .vote_accounts()
+            .into_iter()
+            .filter_map(|(pubkey, (_, account))| {
+                VoteState::from(&account).and_then(|state| {
+                    let timestamp_slot = state.last_timestamp.slot;
+                    if self.slot().checked_sub(timestamp_slot)? <= TIMESTAMP_SLOT_RANGE as u64 {
+                        Some((
+                            pubkey,
+                            (state.last_timestamp.slot, state.last_timestamp.timestamp),
+                        ))
+                    } else {
+                        None
+                    }
+                })
+            })
+            .collect();
+        let slot_duration = Duration::from_nanos(self.ns_per_slot as u64);
+        let epoch = self.epoch_schedule().get_epoch(self.slot());
+        let stakes = HashMap::new();
```

Comment:
I wonder if an empty HashMap implies a heap allocation.  Can we avoid answering that by returning None when epoch_vote_accounts returns None?

Reply:
Ah, yep definitely